kep/20200903-node-resources-allocatable-scoring.md

```diff
@@ -0,0 +1,82 @@
+<!-- /toc -->
+
+## Motivation
+A Kubernetes cluster can consist of nodes with widely varying resource capacities. Likewise, workloads may consist of pods with widely varying resource requests. A pod with high resource requests might only fit on one of the largest capacity nodes. If one or more small pods are scheduled on the largest capacity nodes, the large pod may be unscheduleable. That is a sub-optimal scheduling when there are smaller nodes available that could have taken the small pods. The existing scheduling priorities don't offer a solution to this scenario. To improve cluster utilization, this proposal tries to solve the problem by implementing a scoring plugin ```NodeResourcesAllocatable``` based on the scheduler framework.
+ 
+## Goals
+1. Use a scheduler plugin, which is the most Kubernetes native way, to implement node resources allocatable scoring. 
+2. Provide args to score nodes based on node resources least allocatable or most allocatable.
+3. Provide args to specify different weights for each resource used in the scoring calculations.
+ 
+## Non-Goals
+ 
+## Use Cases
+1. When small and/or large pods may be deployed on a cluster where the large pods can only fit on a subset of nodes (e.g. the largest capacity nodes), the small pods should be scheduled such that there is room, if possible, to schedule the large pods when they are deployed. Otherwise the small pods may land on the large nodes and block the large pods from getting scheduled. If the scheduler uses the ```NodeResourcesAllocatable``` score plugin configured to prioritize least allocatable, that satisfies this use case.
+
+2. Consider a cluster where there are a few permanent large capacity nodes and other smaller capacity ephemeral nodes. The ephemeral nodes may be added to or removed from the cluster based on demand. If unutilized, an ephemeral node may be removed to reduce costs. If we are optimizing to reduce costs, then the scheduler should prioritize placing pods on the large capacity nodes. If the scheduler uses the ```NodeResourcesAllocatable``` score plugin configured to prioritize most allocatable, that satisfies this use case.
+
+## Terms
+
+## Proposal
+
+In order to implement node resources allocatable scoring, we developed a score plugin. The plugin takes an arg ```mode```, which can have values ```Least``` or ```Most```. When ```Least``` is specified nodes with the least allocatable resources are scored higher. When ```Most``` is specified, nodes with the most allocatable resources are scored higher. ```Least``` and ```Most``` are totally opposite semantics, so it's only possible to have one enabled in a scheduler profile. The plugin takes another arg ```resources``` to specify the ```weight``` and ```name``` of each resource used in the scoring calculations.
+
+## Design Details
+We implement a score plugin based on the scheduler framework. Implementation follows a similar pattern to the existing ```NodeResourcesLeastAllocated``` score plugin.
+
```

Comment:
It'd be good to give a concrete example explaining how the score is calculated.

Reply:
Done